extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,270 @@
+            }
+            buffer.put(b, off, remaining);
+            sendPartRequest(useRequest);
+            off += remaining;
+            len -= remaining;
+        }
+    }
+
+    public void flush() throws IOException {
+        final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
+        if (pendingRequests.size() == nextSlotId) {
+            // Nothing to flush
+            return;
+        }
+        final OutgoingRequest request = pendingRequests.get(nextSlotId);
+        if (request.buffer.position() != 0 && request.future == null) {
+            sendPartRequest(request);
+        }
+    }
+
+    /**
+     * Try to finish the multipart upload and close the stream. Cancel the upload if an error occurs.
+     *
+     * @throws IOException if an error occurs while closing the stream
+     */
+    public void close() throws IOException {
+        if (uploadId == null) {
+            return;
+        }
+        try {
+            flush();
+            completeMultipartUpload();
+        } catch (final IOException e) {
+            abort();
+            throw new IOException(String.format("Error closing S3OutputStream for uri %s, aborting upload.", uri), e);
+        }
+        uploadId = null;
+    }
+
+    /**
+     * Abort the multipart upload if it is in progress and close the stream.
+     */
+    void abort() throws IOException {
+        if (uploadId == null) {
+            return;
+        }
+        final AbortMultipartUploadRequest abortRequest = AbortMultipartUploadRequest.builder()
+                .bucket(uri.bucket().orElseThrow())
+                .key(uri.key().orElseThrow())
+                .uploadId(uploadId)
+                .build();
+        try {
+            s3AsyncClient.abortMultipartUpload(abortRequest).get();
+        } catch (final InterruptedException | ExecutionException | CancellationException e) {
+            throw handleS3Exception(e, String.format("aborting multipart upload for uri %s", uri), s3Instructions);
+        }
+        uploadId = null;
+    }
+
+    ////////// Helper methods and classes //////////
+
+    private static class OutgoingRequest {
+        /**
+         * The buffer for this request
+         */
+        private final ByteBuffer buffer;
+
+        /**
+         * The part number for the part to be uploaded
+         */
+        private int partNumber;
+
+        /**
+         * The future for the part upload
+         */
+        private CompletableFuture<UploadPartResponse> future;
+
+        OutgoingRequest(final int partSize) {
+            buffer = ByteBuffer.allocate(partSize);
+            partNumber = INVALID_PART_NUMBER;
+        }
+    }
+
+    private String initiateMultipartUpload() throws IOException {
+        final CreateMultipartUploadRequest createMultipartUploadRequest = CreateMultipartUploadRequest.builder()
+                .bucket(uri.bucket().orElseThrow())
+                .key(uri.key().orElseThrow())
+                .build();
```

Comment:
Noting, there are additional fields we could consider supporting here. Tagging, storage class, encryption, permissions. Doesn't need to be out of the gate, but something to consider.